### PR TITLE
depend: include anaconda path to libpython on linux 

### PR DESCRIPTION
**Pull request by amoh-godwin**
Anaconda is widely used. It is the recommended way to install python and its modules.
PyInstaller works well with it in windows but did not consider searching for the ```libpython``` in the anaconda distribution on Linux which is most probably the case.
So here is a commit that solves the problem. Issue #3885

### CODE REVIEW

**Comment by amoh-godwin** [comment]
The tests failed not because of anything at my end or in my code. Please verify the error for me. I have no idea how to solve this problem.

File: PyInstaller/depend/bindepend.py [hunk 1]
```diff
@@ -875,6 +876,36 @@ def get_python_library_path():
     # Applies only to non Windows platforms.
 
     if is_unix:
+
+        # Search if it is an Anaconda distribution
+        if re.findall('/anaconda', sys.executable):
```
Comment (htgoebel) [inline PyInstaller/depend/bindepend.py]: Please use `compat.is_conda`.

Beside this:
- Using `re` for searching for static strings is much too much overhead.
- Assuming `/` in the path does not work on e.g. Windows.

File: PyInstaller/depend/bindepend.py [hunk 2]
```diff
@@ -875,6 +876,36 @@ def get_python_library_path():
...
+
+            # Relative anaconda path to the lib
+            rel_lib_path = 'lib/libpython3.7m.so.1.0'
```
Comment (htgoebel) [inline PyInstaller/depend/bindepend.py]: Ouch! This is bound to fail for Python 3.6 and 3.8 and if conda changes the compilation options (which will result in a different filename).

Also  os.path.join() should be used here

File: PyInstaller/depend/bindepend.py [hunk 3]
```diff
@@ -875,6 +876,36 @@ def get_python_library_path():
...
+
+            # Split the path of the python executable
+            # to determine what the full anaconda name is
+            # and what paths preceed it.
+            splits = sys.executable.split('/anaconda')
```
Comment (htgoebel) [inline PyInstaller/depend/bindepend.py]: Again this will not work on e.g. Windows.

File: PyInstaller/depend/bindepend.py [hunk 4]
```diff
@@ -875,6 +876,36 @@ def get_python_library_path():
...
+
+            # The preceding path to the anaconda directory
+            pre_path = splits[0]
+
+            # Split the remainder to seperate any other paths
+            back_splits = splits[1].split('/')
```
Comment (htgoebel) [inline PyInstaller/depend/bindepend.py]: Looks like you are basically calculating `sys.exec_prefix`? (Beside of again not working in e.g. Windows.)

Please give an example of the full paths you have and want to receive.

File: PyInstaller/depend/bindepend.py (hunk 1 above)
Comment (amoh-godwin) [reply PyInstaller/depend/bindepend.py]: I actually tried ```usiing compat.is_conda``` but it returns a false value, and I use Anaconda.

File: PyInstaller/depend/bindepend.py (hunk 2 above)
Comment (amoh-godwin) [reply PyInstaller/depend/bindepend.py]: Please I have used the regex to find a file that starts with 'libpython' and ends with 'm.so.1.0.'.
Making it difficult for any other file to make the cut.
```suggestion
            if re.findall('^libpython.*?m.so.1.0$', entry)
```

File: PyInstaller/depend/bindepend.py (hunk 3 above)
Comment (amoh-godwin) [reply PyInstaller/depend/bindepend.py]: This has not been used

File: PyInstaller/depend/bindepend.py (hunk 4 above)
Comment (amoh-godwin) [reply PyInstaller/depend/bindepend.py]: I have used compat.base_prefix instead. Thanks

**Comment by amoh-godwin** [comment]
> 
> 
> Thanks for the update. Sadly this is just tampering with your old code, bringing new issues (e.g. too loose path-matching, needless use of regexp, etc.). I'm not going to merge this, sorry.
> 
> Please implement what I asked you in my first comment:
> 
> > **extent the search-path in`findLibrary`**, around here: 
> >   
> >     
> >       [pyinstaller/PyInstaller/depend/bindepend.py](https://github.com/pyinstaller/pyinstaller/blob/0f31b35fe96de59e1a6faf692340a9ef93492472/PyInstaller/depend/bindepend.py#L769-L779)
> >     
> >     
> >         Lines 769 to 779
> >       in
> >       [0f31b35](/pyinstaller/pyinstaller/commit/0f31b35fe96de59e1a6faf692340a9ef93492472)
> >     
> >     
> >     
> >     
> > 
> >         
> >           
> >            # Look in the known safe paths. 
> >         
> > 
> >         
> >           
> >            if lib is None: 
> >         
> > 
> >         
> >           
> >                # Architecture independent locations. 
> >         
> > 
> >         
> >           
> >                paths = ['/lib', '/usr/lib'] 
> >         
> > 
> >         
> >           
> >                # Architecture dependent locations. 
> >         
> > 
> >         
> >           
> >                arch = compat.architecture() 
> >         
> > 
> >         
> >           
> >                if arch == '32bit': 
> >         
> > 
> >         
> >           
> >                    paths.extend(['/lib32', '/usr/lib32', '/usr/lib/i386-linux-gnu']) 
> >         
> > 
> >         
> >           
> >                else: 
> >         
> > 
> >         
> >           
> >                    paths.extend(['/lib64', '/usr/lib64', '/usr/lib/x86_64-linux-gnu'])
> 
> Extending the search-path is simple, clean and sustainable.

Extending the search-path is not simple for me:
Probably you can help by pointing out exactly where my code should be.

**Comment by amoh-godwin** [comment]
> Other remarks and questions:
> 
>     * If `is_conda` does not work for you, we need to fix it. What is the output of `.../anconda/bin/python -c "import sys; print sys.version"`?

Sorry. My virtual box just shutdown, and I do not know when I will get it running again.

**Comment by htgoebel** [comment]
> There have been no needless use of regexp, do check again. The only place that regexp was used was when matching was done to find the libpython file itself, not anaconda

Yes, and there is no need to use a regex here: The version numner is known, this the string can easily build (as it is done at some other places) and then a simple string-compare would suffice.

> Extending the search-path is not simple for me:

After line 778 add something like this:
```
    if compat.is_conda:
        # prefer the lib included in this conada environment
        paths.insert(0, os.path.join(compat.base_prefix, 'lib')
```

> Sorry. My virtual box just shutdown, and I do not know when I will get it running again.

No problem for me.

File: PyInstaller/depend/bindepend.py (hunk 1 above)
Comment (htgoebel) [reply PyInstaller/depend/bindepend.py]: What is the output of `.../anconda/bin/python -c "import sys; print sys.version"`?

**Comment by htgoebel** [comment]
Superseded by #4015, which solves some of the issues discussed here.

**Pull request closed without merging**